index.js

```diff
@@ -0,0 +1,213 @@
+// 1
+function getDeltaFunction(deltaObject) {
+  return function(subDelta = {}) {
+    const delta = {...deltaObject, ...subDelta}
+    return  delta.b * delta.b - 4 * delta.a * delta.c;
+  }
+}
+
+// delta = b*b - 4*a*c
+const getDelta = getDeltaFunction({ a: 1, b: 2, c: 3 });
+console.log(getDelta()); // -8
+console.log(getDelta({ b: 12 })); // 132
+console.log(getDelta({ a: 4, b: 10 })); // 52
+console.log(getDelta({ a: 4, b: 0 })); // -48
+
+// The nested function returns NaN if any of the arguments are missing
+console.log(getDeltaFunction({ a: 1, c: 3 })()); // NaN
+console.log(getDeltaFunction({ a: 1, c: 3 })({ b: 15 })); // 213
+
+// 2
+function positiveSum(numbers) {
+  let numbersSum = 0;
+
+  numbers.forEach(function(number) {
+    if (number > 0) {
+      numbersSum = numbersSum + number;
+    }
+  })
+
+  return numbersSum;
+}
+
+console.log(positiveSum([1, 5, 10]));
+console.log(positiveSum([-1, 5, -10]));
+
+// 3
+function countSheep(shed) {
+  return shed.filter(Boolean).length;
```

Comment:
Using native functions from JavaScript as callbacks can cause unexpected results. A good example is `parseInt`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

```
const numbersAsStrings = ['1', '10', '20'];

const numbers = numbersAsStrings.map(parseInt);

console.log(numbers); // [1, NaN, NaN]
```

It's because the callback passed to the `map` function receives **three** arguments:
- the current element from the array
- the current index
- the array the `map()` was called upon

Using `numbersAsStrings.map(parseInt)` is equal to this:
```
numbersAsStrings.map(function(number, index, array) {
  return parseInt(number, index, array);
});
```

This causes a problem - the `parseInt` actually supports two arguments with the second being the `radix`.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

This causes a problem, and strings are parsed to `NaN`. We could fix it the following way:

```
const numbers = numbersAsStrings.map(function(number) {
  // ignoring index and array
  return parseInt(number);
});
```

In your particular case, the `Boolean` function accepts just one parameter so there is no bug in your code. However, because of issues like that, I suggest avoiding passing native JavaScript functions as callbacks, and instead do
```
shed.filter(function(shedElement) {
  return Boolean(shedElement);
})
```

Reply:
ok, didn't know that - fixed